// @nestjs/ng-universal/dist/utils/setup-universal.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCacheOptions = exports.setupUniversal = void 0;
const express_engine_1 = require("@nguniversal/express-engine");
const express = require("express");
const in_memory_cache_storage_1 = require("../cache/in-memory-cache.storage");
const cahce_key_by_original_url_generator_1 = require("../cache/cahce-key-by-original-url.generator");
const DEFAULT_CACHE_EXPIRATION_TIME = 60000;
function setupUniversal(app, ngOptions) {
    const cacheOptions = getCacheOptions(ngOptions);
    app.engine('html', (_, options, callback) => {
        let cacheKey;
        if (cacheOptions.isEnabled) {
            const cacheKeyGenerator = cacheOptions.keyGenerator;
            cacheKey = cacheKeyGenerator.generateCacheKey(options.req);
            const cacheHtml = cacheOptions.storage.get(cacheKey);
            if (cacheHtml) {
                return callback(null, cacheHtml);
            }
        }
        express_engine_1.ngExpressEngine({
            bootstrap: ngOptions.bootstrap,
            providers: [
                {
                    provide: 'serverUrl',
                    useValue: `${options.req.protocol}://${options.req.get('host')}`
                },
                ...(ngOptions.extraProviders || [])
            ]
        })(_, options, (err, html) => {
            if (cacheOptions.isEnabled && cacheKey) {
                cacheOptions.storage.set(cacheKey, html, cacheOptions.expiresIn);
            }
            callback(null, html);
        });
    });
    app.set('view engine', 'html');
    app.set('views', ngOptions.viewsPath);
    app.get(ngOptions.rootStaticPath, express.static(ngOptions.viewsPath, {
        maxAge: 600
    }));
}
exports.setupUniversal = setupUniversal;
function getCacheOptions(ngOptions) {
    if (!ngOptions.cache) {
        return {
            isEnabled: false
        };
    }
    if (typeof ngOptions.cache !== 'object') {
        return {
            isEnabled: true,
            storage: new in_memory_cache_storage_1.InMemoryCacheStorage(),
            expiresIn: DEFAULT_CACHE_EXPIRATION_TIME,
            keyGenerator: new cahce_key_by_original_url_generator_1.CacheKeyByOriginalUrlGenerator()
        };
    }
    return {
        isEnabled: true,
        storage: ngOptions.cache.storage || new in_memory_cache_storage_1.InMemoryCacheStorage(),
        expiresIn: ngOptions.cache.expiresIn || DEFAULT_CACHE_EXPIRATION_TIME,
        keyGenerator: ngOptions.cache.keyGenerator || new cahce_key_by_original_url_generator_1.CacheKeyByOriginalUrlGenerator()
    };
}
exports.getCacheOptions = getCacheOptions;
